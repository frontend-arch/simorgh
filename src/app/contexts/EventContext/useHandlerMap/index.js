/**
 * Will store mappings between query selectors and event handler functions
 * @param {object} _map a datastore for mapping query selectors to array of event handler functions
 */
const useHandlerMap = _map => e => {
  if (e && e.target) {
    const attrs = Object.keys(_map).filter(attr => e.target.matches(attr));

    attrs.forEach(attr => {
      const handlers = _map[attr] || [];
      handlers.forEach(handler => {
        if (typeof handler === 'function') {
          handler(e);
        }
      });
    });
  }
};

export default useHandlerMap;
