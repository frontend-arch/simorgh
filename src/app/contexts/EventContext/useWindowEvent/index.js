const useWindowEvent = (event, handler, passive = false) => () => {
  window.addEventListener(event, handler, passive);
  const cleanup = () => {
    window.removeEventListener(event, handler, passive);
  };
  return cleanup;
};

export default useWindowEvent;
