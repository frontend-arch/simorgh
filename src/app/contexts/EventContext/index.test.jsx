import React from 'react';
import { render } from '@testing-library/react';
import { func } from 'prop-types';
import * as useWindowEvent from './useWindowEvent';
import * as useHandlerMap from './useHandlerMap';

const useContextSpy = jest.spyOn(React, 'useContext');
const useStateSpy = jest.spyOn(React, 'useState');
const useEffectSpy = jest.spyOn(React, 'useEffect');
const useWindowEventSpy = jest.spyOn(useWindowEvent, 'default');
const useHandlerMapSpy = jest.spyOn(useHandlerMap, 'default');

const { EventContextProvider, EventContext } = require('./index');

const Component = () => {
  React.useContext(EventContext);
  return null;
};

afterEach(() => {
  jest.clearAllMocks();
});

describe('EventContext', () => {
  it('should call hooks', () => {
    render(
      <EventContextProvider>
        <Component />
      </EventContextProvider>,
    );

    expect(useWindowEventSpy).toHaveBeenCalled();

    expect(useHandlerMapSpy).toHaveBeenCalled();

    expect(useContextSpy).toHaveBeenCalledWith(EventContext);

    expect(useStateSpy).toHaveBeenCalledWith({});
  });

  describe('useClickTracker', () => {
    const ClickComponent = ({ handler }) => {
      const { useClickTracker, handlerMap } = React.useContext(EventContext);
      useClickTracker('my-test-component', handler);
      return (
        <ul>
          {Object.entries(handlerMap).map(([key, values]) => (
            <li key={key}>
              {key}:{values.map(v => v.toString()).join('\n')}
            </li>
          ))}
        </ul>
      );
    };

    ClickComponent.propTypes = {
      handler: func.isRequired,
    };

    it('should useEffect and match snapshot', () => {
      const handler = jest.fn();
      const { container } = render(
        <EventContextProvider>
          <ClickComponent handler={handler} />
        </EventContextProvider>,
      );

      expect(useEffectSpy).toHaveBeenCalled();
      expect(container).toMatchSnapshot();
      expect(handler).not.toHaveBeenCalled();
    });

    it('should call handler', () => {
      const handler = jest.fn();
      const { container } = render(
        <EventContextProvider>
          <ClickComponent handler={handler} />
        </EventContextProvider>,
      );

      window.matches = selector => ['my-test-component'].includes(selector);
      window.dispatchEvent(new Event('click'));

      expect(useEffectSpy).toHaveBeenCalled();
      expect(container).toMatchSnapshot();
      expect(handler).toHaveBeenCalled();
    });
  });
});
