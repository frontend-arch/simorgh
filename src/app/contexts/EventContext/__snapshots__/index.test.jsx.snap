// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventContext useClickTracker should call handler 1`] = `
<div>
  <ul>
    <li>
      my-test-component
      :
      function mockConstructor() {
        return fn.apply(this, arguments);
      }
    </li>
  </ul>
</div>
`;

exports[`EventContext useClickTracker should useEffect and match snapshot 1`] = `
<div>
  <ul>
    <li>
      my-test-component
      :
      function mockConstructor() {
        return fn.apply(this, arguments);
      }
    </li>
  </ul>
</div>
`;
